api/v1beta2/auth_config_conversion.go

```diff
@@ -0,0 +1,1032 @@
+			dst.Spec.Response.Success.Headers = make(map[string]HeaderSuccessResponseSpec)
+		}
+		name, response := convertSuccessResponseFrom(responseSrc)
+		dst.Spec.Response.Success.Headers[name] = HeaderSuccessResponseSpec{
+			SuccessResponseSpec: response,
+		}
+	}
+
+	for _, responseSrc := range src.Spec.Response {
+		if responseSrc.Wrapper != "envoyDynamicMetadata" {
+			continue
+		}
+		if dst.Spec.Response.Success.DynamicMetadata == nil {
+			dst.Spec.Response.Success.DynamicMetadata = make(map[string]SuccessResponseSpec)
+		}
+		name, response := convertSuccessResponseFrom(responseSrc)
+		dst.Spec.Response.Success.DynamicMetadata[name] = response
+	}
+
+	// callbacks
+	if src.Spec.Callbacks != nil {
+		dst.Spec.Callbacks = make(map[string]CallbackSpec, len(src.Spec.Callbacks))
+		for _, callbackSrc := range src.Spec.Callbacks {
+			name, callback := convertCallbackFrom(callbackSrc)
+			dst.Spec.Callbacks[name] = callback
+		}
+	}
+
+	// status
+	dst.Status = convertStatusFrom(src.Status)
+
+	logger.V(1).Info("finished converting resource", "dst", dst)
+
+	return nil
+}
+
+func convertPatternExpressionTo(src PatternExpression) v1beta1.JSONPatternExpression {
+	return v1beta1.JSONPatternExpression{
+		Selector: src.Selector,
+		Operator: v1beta1.JSONPatternOperator(src.Operator),
+		Value:    src.Value,
+	}
+}
+
+func convertPatternExpressionFrom(src v1beta1.JSONPatternExpression) PatternExpression {
+	return PatternExpression{
+		Selector: src.Selector,
+		Operator: PatternExpressionOperator(src.Operator),
+		Value:    src.Value,
+	}
+}
+
+func convertPatternExpressionOrRefTo(src PatternExpressionOrRef) v1beta1.JSONPattern {
+	return v1beta1.JSONPattern{
+		JSONPatternExpression: convertPatternExpressionTo(src.PatternExpression),
+		JSONPatternRef: v1beta1.JSONPatternRef{
+			JSONPatternName: src.PatternRef.Name,
+		},
+	}
+}
+
+func convertPatternExpressionOrRefFrom(src v1beta1.JSONPattern) PatternExpressionOrRef {
+	return PatternExpressionOrRef{
+		PatternExpression: convertPatternExpressionFrom(src.JSONPatternExpression),
+		PatternRef: PatternRef{
+			Name: src.JSONPatternRef.JSONPatternName,
+		},
+	}
+}
+
+func convertEvaluatorCachingTo(src *EvaluatorCaching) *v1beta1.EvaluatorCaching {
+	if src == nil {
+		return nil
+	}
+	return &v1beta1.EvaluatorCaching{
+		Key: convertValueOrSelectorTo(src.Key),
+		TTL: src.TTL,
+	}
+}
+
+func convertEvaluatorCachingFrom(src *v1beta1.EvaluatorCaching) *EvaluatorCaching {
+	if src == nil {
+		return nil
+	}
+	return &EvaluatorCaching{
+		Key: convertValueOrSelectorFrom(src.Key),
+		TTL: src.TTL,
+	}
+}
+
+func convertValueOrSelectorTo(src ValueOrSelector) v1beta1.StaticOrDynamicValue {
+	return v1beta1.StaticOrDynamicValue{
+		Value:     string(src.Value.Raw),
+		ValueFrom: convertSelectorTo(src),
+	}
+}
+
+func convertValueOrSelectorFrom(src v1beta1.StaticOrDynamicValue) ValueOrSelector {
+	value := k8sruntime.RawExtension{}
+	if src.ValueFrom.AuthJSON == "" {
+		value.Raw = []byte(src.Value)
+	}
```

Comment:
```suggestion
	if src.ValueFrom.AuthJSON != "" {
		value.Raw = []byte(src.Value)
	}
```
I think this should be `!=` 🤔

Reply:
Not really. Notice that, when `ValueFrom.AuthJSON` is empty, it sets the value from `Value`.

These fields are the _exactly one-of_ kind (i.e. mutually exclusive), nevertheless typed in Golang as `string`; not as pointers. Therefore the distinction between empty value and null value has to be semantic.

`ValueFrom.AuthJSON` is expected to contain a path within the Authorization JSON; if it's empty, then there is no way to select anything and the user must have meant a null `ValueFrom.AuthJSON`, and value set from `Value`.  If `Value` is also empty, then the user must have meant indeed the empty string as value.

Reply:
Oh I see, thanks for the explaination ! That makes sense 👍